crypto/x509/x509_vpm.c

```diff
@@ -313,8 +321,50 @@ int X509_VERIFY_PARAM_set1(X509_VERIFY_PARAM *to,
   return ret;
 }
 
-static int int_x509_param_set1(char **pdest, size_t *pdestlen, const char *src,
-                               size_t srclen) {
+static int int_x509_param_set1_email(char **pdest, size_t *pdestlen,
+                                     const char *src, size_t srclen) {
+  void *tmp;
+  if (src != NULL) {
+    // Setting |srclen| to 0 to automatically detect the length of |src| is an
+    // OpenSSL quirk that AWS-LC isn't keen on supporting. However, consumers
+    // often assume OpenSSL semantics from AWS-LC, so it's supported in this
+    // case.
+    if (srclen == 0) {
+      srclen = strlen(src);
+    }
+
+    tmp = OPENSSL_memdup(src, srclen);
+    if (tmp == NULL) {
+      return 0;
+    }
+  } else {
+    // This allows an empty string to disable previously configured checks.
+    // This is an OpenSSL quirk that AWS-LC isn't keen on supporting. However,
+    // consumers often assume OpenSSL semantics from AWS-LC, so it's supported
+    // in this case.
+    tmp = NULL;
+    srclen = 0;
+  }
+
+  if (*pdest != NULL) {
+    OPENSSL_free(*pdest);
+  }
+  *pdest = tmp;
+  if (pdestlen != NULL) {
+    *pdestlen = srclen;
+  }
+  return 1;
+}
+
+// IP addresses work slightly differently, so we use another function to
+// differentiate from emails. |X509_VERIFY_PARAM_set1_ip| takes a const
+// unsigned char*, instead of a const char*, so the same strlen logic that was
+// being used is not quite suitable here.
+// We keep the original behavior that BoringSSL left, but only for IP addresses.
+// We can align the behavior with |int_x509_param_set1_email| like OpenSSL has
+// been doing if needed.
```

Comment:
How hard is this to do now? This will lead to a confusing experience with different functions treating 0 in different ways.

Reply:
`X509_VERIFY_PARAM_set1_ip` takes a `const unsigned char*` with the `ip` parameter, instead of a `const char*`. OpenSSL then attempts to transform the `unsigned char` pointer into an `const char` and passes it into `int_x509_param_set1`. `int_x509_param_set1` calls `strlen` on the pointer, but this results in a fatal failure if the  `const unsigned char*` pointer is non-NULL terminated.

* https://github.com/openssl/openssl/blob/7197abddb891933f52ec84dafb41b685d4a1d122/crypto/x509/x509_vpm.c#L466-L467
* https://github.com/openssl/openssl/blob/7197abddb891933f52ec84dafb41b685d4a1d122/crypto/x509/x509_vpm.c#L243

I chose to keep the original behavior since it's effectively getting around the weird behavior OpenSSL was doing. Consumers of `X509_VERIFY_PARAM_set1_ip` were likely defining the length since using 0 would often result in fatal failures, unless they were manually appending `\0` at the end of the IP string.
We could do this now, but it would easily expose the obvious fatal failures when misusing `X509_VERIFY_PARAM_set1_ip`.